### PR TITLE
feat(subsystems): add creating and loading `reflections.cache` for subsystems

### PR DESCRIPTION
**Pull request by DarkWeird**
### Contains

Creating and loading `reflection.cache` for subsystems.
define upToDate of `cacheReflection` task for engine, modules, subsystems.

### How to test

1. Open Terasology
2. Check that DiscordRPCSystem works (should connects to Discord App, if enabled)

### CODE REVIEW

**Comment by iHDeveloper** [comment]
Can confirm the subsystem works!

Now the Discord-RPC changes when the player is playing (not in the lobby aka main menu)
image

**Comment by jdrueckert** [comment]
@iHDeveloper Where does that "Solo | Test" come from in the screenshot? 🤔

**Comment by DarkWeird** [comment]
> @iHDeveloper Where does that "Solo | Test" come from in the screenshot? thinking

There from `DiscordRPCSystem#getGame`
https://github.com/MovingBlocks/Terasology/blob/04572ef4e74259f4b47650ff9a85360c804f8287/subsystems/DiscordRPC/src/main/java/org/terasology/engine/subsystem/discordrpc/DiscordRPCSystem.java#L35-L44

**Comment by jdrueckert** [comment]
> There from `DiscordRPCSystem#getGame`
> 
> https://github.com/MovingBlocks/Terasology/blob/04572ef4e74259f4b47650ff9a85360c804f8287/subsystems/DiscordRPC/src/main/java/org/terasology/engine/subsystem/discordrpc/DiscordRPCSystem.java#L35-L44

I see it differentiates between "Hosting" and "Singleplayer"... What happens for players joining a hosted game? Will `networkMode` `DEDICATED_SERVER`, `NONE` or something entirely different?

**Comment by DarkWeird** [comment]
> > There from `DiscordRPCSystem#getGame`
> > 
> > https://github.com/MovingBlocks/Terasology/blob/04572ef4e74259f4b47650ff9a85360c804f8287/subsystems/DiscordRPC/src/main/java/org/terasology/engine/subsystem/discordrpc/DiscordRPCSystem.java#L35-L44
> 
> I see it differentiates between "Hosting" and "Singleplayer"... What happens for players joining a hosted game? Will `networkMode` `DEDICATED_SERVER`, `NONE` or something entirely different?

Will "Playing online"

File: subsystems/build.gradle [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+            // Without the .mkdirs() we might hit a scenario where the classes dir doesn't exist yet
+            Reflections reflections = new Reflections(new ConfigurationBuilder()
+                    .filterInputsBy { it.startsWith('org/terasology/')}
+                    .addUrls(inputs.files.collect { it.toURI().toURL() })
+                    .setScanners(new TypeAnnotationsScanner(), new SubTypesScanner()))
+            reflections.save("$outputDir/reflections.cache")
+        }
+    }
+    tasks.getByName('jar').dependsOn('cacheReflections')
+}
```
Comment (jdrueckert) [inline subsystems/build.gradle:48]:
```suggestion
}

```

**Review by jdrueckert — APPROVED** [review]

**Comment by iHDeveloper** [comment]
> @iHDeveloper Where does that "Solo | Test" come from in the screenshot? 🤔

That's the game state of the player that represents what's the player playing.

The `DiscordRPCSystem` changes the state of the player's discord RPC as follows:

`Solo | GameName` meaning the player is currently playing a single-player game
`Hosting | GameName` meaning the player is hosting a game
`Playing Online` meaning the player is playing online in a server (not hosting it)

The state doesn't represent a party state and it's confusing more than encouraging the player's friends to join.
I'm planning to improve the state in #4271 by following the rich presence best practices

> The data in your players’ profiles is the first thing that others on Discord will see about your game, both those familiar with it and those who have never played. It should answer two questions: can I play with my friend right now, and if not, when can I? Show data like:
> 
> - What the player is currently doing
> - How much time has elapsed or remains (if applicable)
> - Their party state
> - Your cool artwork!
>
> _How should you think about the data you show? -Rich Presence Best Practices_

_[Source: **Discord Developer Documentation**](https://discord.com/developers/docs/rich-presence/best-practices#how-should-you-think-about-the-data-you-show)_

**Comment by jdrueckert** [comment]
@iHDeveloper Thanks for the detailed information 🙂 
I believe, Solo, Hosting and Playing Online describe different dimensions and do not cover the full space.
For instance, the opposite of solo is not hosting or playing online in general, it's multiplayer (can also be played on LAN which is not necessarily online).

I don't think this is critical and getting it to work at all is good progress, so I merged this PR. However, I believe we should consider doing a follow-up that improves the naming of the states.

**Comment by keturn** [comment]
I hope I remember to come back and look at this again later.

I've got no objections to this having been merged, but the reflections cache stuff is the part of our gradle build that has the most problems fitting in with gradle's default build logic. For the benefit of my own understanding and working on things like #4022, I'll want to learn what's been added to the subsystem build here and how it is same/different from what happens for the rest of engine and the plugins.

ping #4015 

(but also also I was hoping to get gestalt v7 merged before returning to #4022, and between that and some other work-in-progress we might not need to worry so much about reflections.cache anymore?)

**Comment by DarkWeird** [comment]
> I hope I remember to come back and look at this again later.
> 
> I've got no objections to this having been merged, but the reflections cache stuff is the part of our gradle build that has the most problems fitting in with gradle's default build logic. For the benefit of my own understanding and working on things like #4022, I'll want to learn what's been added to the subsystem build here and how it is same/different from what happens for the rest of engine and the plugins.
> 
> ping #4015 
> 
> (but also also I was hoping to get gestalt v7 merged before returning to #4022, and between that and some other work-in-progress we might not need to worry so much about reflections.cache anymore?)

Gestaltv7 using reflection.cache too (only in json format, not xml)

Better solution - use annotation processing like ClassIndex. Common for gradle and idea.
Also we have specific build directories...because gestalt-module's SourceModule...

**Pull request merged**